src/coreclr/jit/optimizer.cpp

```diff
@@ -7608,17 +7608,805 @@ bool Compiler::optIsRangeCheckRemovable(GenTree* tree)
+
+        noway_assert(b1->bbJumpKind == BBJ_RETURN);
+        noway_assert(b2->bbJumpKind == BBJ_RETURN);
+        noway_assert(b1->bbNext == b2);
+        noway_assert(b2->bbNext == b3);
+        noway_assert(b3);
+    }
+    else
+    {
+        noway_assert(b1->bbJumpKind == BBJ_COND);
+        noway_assert(b2->bbJumpKind == BBJ_COND);
+        noway_assert(b1->bbJumpDest == b2->bbJumpDest);
+        noway_assert(b1->bbNext == b2);
+        noway_assert(b2->bbNext);
+    }
+
+    // Get rid of the second block
+
+    comp->fgUnlinkBlock(b2);
+    b2->bbFlags |= BBF_REMOVED;
+    // If b2 was the last block of a try or handler, update the EH table.
+    comp->ehUpdateForDeletedBlock(b2);
+
+    if (optReturnBlock)
+    {
+        // Get rid of the third block
+        comp->fgUnlinkBlock(b3);
+        b3->bbFlags |= BBF_REMOVED;
+        // If b3 was the last block of a try or handler, update the EH table.
+        comp->ehUpdateForDeletedBlock(b3);
+    }
+
+    if (!optReturnBlock)
+    {
+        // Update bbRefs and bbPreds
+        //
+        // Replace pred 'b2' for 'b2->bbNext' with 'b1'
+        // Remove  pred 'b2' for 'b2->bbJumpDest'
+        comp->fgReplacePred(b2->bbNext, b2, b1);
+        comp->fgRemoveRefPred(b2->bbJumpDest, b2);
+    }
+
+    // Update loop table
+    comp->fgUpdateLoopsAfterCompacting(b1, b2);
+    if (optReturnBlock)
+    {
+        comp->fgUpdateLoopsAfterCompacting(b1, b3);
+    }
+}
+
+//
+//  optOptimizeBoolsReturnBlock: Optimize boolean when b1.bbJumpKind is BBJ_COND and b2.bbJumpKind is BBJ_RETURN
+//
+//  Arguments:
+//      comp    The instance of Compiler class
+//
+//  Returns:
+//      changeReturn    Set to true if boolean optimization is done and b1, b2 and b3 are folded into b1.
+//
+//  Notes:
+//      Below members of OptBoolsDsc are set when called:
+//          b1      The first Basic Block with the BBJ_COND conditional jump type
+//          b2      The next basic block of b1 with the BBJ_RETURN
+//          b3      The next basic block of b2 with GT_RETURN
+//
+//      For example, (x==0 && y==0) generates:
+//          b1: GT_JTRUE (BBJ_COND)
+//          b2: GT_RETURN (BBJ_RETURN)
+//          b3: GT_RETURN (BBJ_RETURN),
+//      and it is folded into
+//          b1: GT_RETURN (BBJ_RETURN)
+//
+//      if b1.bbJumpDest == b2.bbNext, it transforms
+//          b1 : brtrue(t1, b3)
+//          b2 : ret(t2)
+//          b3 : ret(0)
+//      to
+//          b1 : ret((!t1) && t2)
+//
+bool OptBoolsDsc::optOptimizeBoolsReturnBlock(Compiler* comp)
+{
+    assert(b1 != nullptr && b2 != nullptr && b3 != nullptr);
+
+    bool changeReturn = false;
+
+    sameTarget          = false;
+    Statement* const s1 = optOptimizeBoolsChkBlkCond();
+    if (s1 == nullptr)
+    {
+        return changeReturn;
+    }
+
+    // Find the branch conditions of b1 and b2
+
+    GenTree* const op1 = optIsBoolComp(comp, &t1);
+    if (op1 == nullptr)
+    {
+        return changeReturn;
+    }
+    c1 = op1;
```

Comment:
```suggestion
    m_c1 = optIsBoolComp(comp, &t1);
    if (m_c1 == nullptr)
    {
        return changeReturn;
    }
```

and similar below. Why use a temp?

Reply:
Done.